components/script/dom/htmlelement.rs

```diff
@@ -897,6 +937,82 @@ impl HTMLElement {
+        // Step 1: Let fragment be a new DocumentFragment whose node document is document.
+        let document = document_from_node(self);
+        let fragment = DocumentFragment::new(&document);
+
+        // Step 2: Let position be a position variable for input, initially pointing at the start
+        // of input.
+        let mut position = input.chars().peekable();
+
+        // Step : Let text be the empty string.
+        let mut text = String::new();
+
+        // Step 4
+        while let Some(ch) = position.next() {
+            match ch {
+                // While position is not past the end of input, and the code point at position is
+                // either U+000A LF or U+000D CR:
+                '\u{000A}' | '\u{000D}' => {
+                    if ch == '\u{000D}' && position.peek() == Some(&'\u{000A}') {
+                        // a \r\n pair should only generate one <br>,
+                        // so just skip the \r.
+                        position.next();
+                    }
+
+                    if !text.is_empty() {
+                        append_text_node_to_fragment(&document, &fragment, text);
+                        text = String::new();
+                    }
+
+                    let br = HTMLBRElement::new(local_name!("br"), None, &document, None);
+                    fragment.upcast::<Node>().AppendChild(br.upcast()).unwrap();
+                },
+                _ => {
+                    // Collect a sequence of code points that are not U+000A LF or U+000D CR from
+                    // input given position, and set text to the result.
+                    text.push(ch);
+                },
+            }
+        }
+
+        // If text is not the empty string, then append a new Text node whose data is text and node
+        // document is document to fragment.
+        if !text.is_empty() {
+            append_text_node_to_fragment(&document, &fragment, text);
+        }
+
+        fragment
+    }
+
+    /// <https://html.spec.whatwg.org/multipage/#merge-with-the-next-text-node>
+    fn merge_with_the_next_text_node(&self, node: DomRoot<Node>) {
+        // Step 1: Let next be node's next sibling.
+        if let Some(next) = node.GetNextSibling() {
+            // Step 2: If next is not a Text node, then return.
+            if next.is::<Text>() {
+                // Step 3: Replace data with node, node's data's length, 0, and next's data.
+                let node_chars = node.downcast::<CharacterData>().unwrap();
+                let next_chars = next.downcast::<CharacterData>().unwrap();
+
+                if let Err(error) =
+                    node_chars.ReplaceData(node_chars.Length(), 0, next_chars.Data())
+                {
+                    warn!("Failed to replace character data: {:?}", error);
+
+                    return;
+                }
+
+                // Step 4: If next's parent is non-null, then remove next.
```

Comment:
I don't see the parent condition in the spec, do you need it?

Reply:
Can probably be removed, since remove_self already makes sure there's a parent